text/0000-bee-network/0000-bee-network.md

```diff
@@ -0,0 +1,295 @@
+    ```
+
+    Note, that the last function is `async`. That is because of possible delays when performing domain name resolution.
+
+* `Protocol`: an `enum`eration of supported communication protocols:
+
+    ```rust
+    #[non_exhaustive]
+    pub enum Protocol {
+        Tcp,
+    }
+    ```
+
+    Note, that future updates may support different protocols, which is why this enum is declared as `non_exhaustive`, see `Unresolved Questions` section.
+
+* `Url`: an `enum`eration that can always be constructed from an `Address` and a `Protocol` (infallible), or from a `&str`, which can fail if parsing or domain name resolution fails. If successfull however, it resolves to an Ipv4 or Ipv6 address stored in variant of the `enum` depending on the url scheme part. Note, that this crate therefore expects the Url string to always provide a scheme (e.g. `tcp://`) and a port (e.g. `15600`) when specifying an endpoint's address.
+
+    ```rust
+    #[non_exhaustive]
+    enum Url {
+        // Represents the address of an endpoint communicating over TCP.
+        Tcp(Address),
+    }
+    ```
+
+    Note, that future updates may support different protocols, which is why this enum is declared as `non_exhaustive`, see `Unresolved Questions` section.
+
+    ```rust
+    // Infallible construction of a `Url` from an `Address` and a `Protocol`.
+    fn new(addr: Address, proto: Protocol) -> Self { ... }
+
+    // Fallible construction of a `Url` from `&str`.
+    async fn from_url_str(url: &str) -> Result<Self> { ... }
+    ```
+
+    Note, that the second function is `async`. That is because of possible delays when performing domain name resolution.
+
+
+
+## `Endpoint` and `EndpointId`
+
+To model the remote part of a connection, this crate introduces the `Endpoint` type:
+
+```rust
+struct Endpoint {
+    // The id of the endpoint.
+    id: EndpointId,
+
+    // The address of the endpoint.
+    address: Address,
+
+    // The protocol used to communicate with that endpoint.
+    protocol: Protocol,
+}
+```
+
+Note: In a peer-to-peer (p2p) network peers usually have two endpoints initially as they are actively trying to connect, but also need to accept connections from their peers. This crate is agnostic about how to handle duplicate connections as this is usually resolved by a handshaking protocol defined at a higher layer.
+
+To uniquely identify an `Endpoint`, this crate proposes the `EndpointId` type, which can be implemented as a wrapper around an `Address`.
+
+```rust
+#[derive(Clone, Copy, Eq, PartialEq, Hash)]
+struct EndpointId(Address);
+```
+
+Note, that this `EndpointId` should be `Clone` and `Copy`, and requires to implement `Eq`, `PartialEq`, and `Hash`, because it is used as a hashmap key in several instances.
+
+## `Command`
+
+`Command`s are messages that are supposed to be issued by higher level business logic like the protocol layer. They are implemented as an `enum`, which in Rust is one way of expressing a polymorphic type:
```

Comment:
```suggestion
`Command`s are messages that are supposed to be issued by higher level business logic like a protocol layer. They are implemented as an `enum`, which in Rust is one way of expressing a polymorphic type:
```

A bit subtle but it ties this layer a bit less to *our* protocol.

Reply:
Changed.